lib/api/src/grpc/transport_channel_pool.rs

```diff
@@ -35,35 +89,72 @@ impl TransportChannelPool {
-                for _ in 0..self.pool_size {
-                    let channel = Self::make_channel(self.grpc_timeout, uri.clone()).await?;
-                    channels.push(channel);
-                }
-                let result = channels[0].clone();
-                guard.insert(uri.clone(), channels);
-                Ok(result)
+                let channels =
+                    ChannelsToUri::init(uri.clone(), self.pool_size, self.grpc_timeout).await?;
+                let channel = channels.choose();
+                guard.insert(uri, channels);
+                Ok(channel)
             }
-            Some(channels) => Ok(channels[0].clone()),
+            Some(channels) => Ok(channels.choose()),
         }
     }
 
-    pub async fn get_pooled_channel(&self, uri: &Uri) -> Option<Channel> {
+    async fn drop_pool(&self, uri: &Uri) {
+        let mut guard = self.uri_to_pool.write().await;
+        guard.remove(uri);
+    }
+
+    async fn get_pooled_channel(&self, uri: &Uri) -> Option<Channel> {
         let guard = self.uri_to_pool.read().await;
-        guard
-            .get(uri)
-            .and_then(|channels| channels.choose(&mut rand::thread_rng()))
-            .cloned()
+        guard.get(uri).map(|channels| channels.choose())
     }
 
-    pub async fn get_or_create_pooled_channel(&self, uri: &Uri) -> Result<Channel, Error> {
+    async fn get_or_create_pooled_channel(&self, uri: &Uri) -> Result<Channel, TonicError> {
         match self.get_pooled_channel(uri).await {
             None => self.init_pool_for_uri(uri.clone()).await,
             Some(channel) => Ok(channel),
         }
     }
+
+    async fn get_last_created(&self, uri: &Uri) -> Option<Instant> {
+        let guard = self.uri_to_pool.read().await;
+        guard.get(uri).map(|channels| channels.init_at)
+    }
+
+    // Allows to use channel to `uri`. If there is no channels to specified uri - they will be created.
+    pub async fn with_channel<T, O: Future<Output = Result<T, Status>>>(
+        &self,
+        uri: &Uri,
+        f: impl Fn(Channel) -> O,
+    ) -> Result<T, RequestError<Status>> {
+        let channel = self.get_or_create_pooled_channel(uri).await?;
+
+        let result = f(channel).await;
+
+        // Reconnect on failure to handle the case with domain name change.
+        match result {
+            Ok(res) => Ok(res),
+            Err(err) => match err.code() {
+                Code::Internal | Code::Unavailable => {
+                    let channel_uptime = Instant::now().duration_since(
+                        self.get_last_created(uri)
+                            .await
+                            .unwrap_or_else(Instant::now),
+                    );
+                    if channel_uptime > CHANNEL_TTL {
+                        self.drop_pool(uri).await;
+                        let channel = self.get_or_create_pooled_channel(uri).await?;
+                        f(channel).await.map_err(RequestError::FromClosure)
```

Comment:
Here we make another attempt to make a request, if
- status code indicates that something is wrong with connection / service
- we created connection long ago